socs/agents/ups/agent.py

```diff
@@ -0,0 +1,341 @@
+        oid_value = oid_value.decode("utf-8")
+
+    # I don't expect any other types at the moment, but just in case.
+    if not isinstance(oid_value, (int, bytes, str)):
+        oid_value = None
+
+    return field_name, oid_value, oid_description
+
+
+def _build_message(get_result, time):
+    """Build the message for publication on an OCS Feed.
+
+    Parameters
+    ----------
+    get_result : pysnmp.smi.rfc1902.ObjectType
+        Result from a pysnmp GET command.
+    time : float
+        Timestamp for when the SNMP GET was issued.
+
+    Returns
+    -------
+    message : dict
+        OCS Feed formatted message for publishing
+    """
+    message = {
+        'block_name': 'ups',
+        'timestamp': time,
+        'data': {}
+    }
+
+    for item in get_result:
+        field_name, oid_value, oid_description = _extract_oid_field_and_value(item)
+
+        if oid_value is None:
+            continue
+
+        message['data'][field_name] = oid_value
+        message['data'][field_name + "_description"] = oid_description
+
+    return message
+
+
+def update_cache(get_result, timestamp):
+    """Update the OID Value Cache.
+
+    The OID Value Cache is used to store each unique OID and will be passed to
+    session.data
+
+    The cache consists of a dictionary, with the unique OIDs as keys, and
+    another dictionary as the value. Each of these nested dictionaries contains the
+    OID values, name, and description (decoded string). An example for a single OID::
+        {"upsIdentModel":
+            {"status": 0,
+                "description": "on"}}
+    Additionally there is connection status and timestamp information under::
+        {"ups_connection":
+            {"last_attempt": 1598543359.6326838,
+            "connected": True}
+        {"timestamp": 1656085022.680916}
+
+    Parameters
+    ----------
+    get_result : pysnmp.smi.rfc1902.ObjectType
+        Result from a pysnmp GET command.
+    timestamp : float
+        Timestamp for when the SNMP GET was issued.
+    """
+    oid_cache = {}
+    try:
+        for item in get_result:
+            field_name, oid_value, oid_description = _extract_oid_field_and_value(item)
+            if oid_value is None:
+                continue
+
+            # Update OID Cache for session.data
+            oid_cache[field_name] = {"status": oid_value}
+            oid_cache[field_name]["description"] = oid_description
+            oid_cache['ups_connection'] = {'last_attempt': time.time(),
+                                           'connected': True}
+            oid_cache['timestamp'] = timestamp
+    # This is a TypeError due to nothing coming back from the yield,
+    # so get_result is None here and can't be iterated.
+    except TypeError:
+        oid_cache['ups_connection'] = {'last_attempt': time.time(),
+                                       'connected': False}
+        raise ConnectionError('No SNMP response. Check your connection.')
+
+    return oid_cache
```

Comment:
Yeah, this is definitely doable for the ibootbar and ups agents. The M1000 would need some work because of the timings. I can look into starting this sometime.